checker/domain.go

```diff
@@ -22,14 +22,36 @@ type DomainStatus int32
 
 // In order of precedence.
 const (
-	DomainSuccess           DomainStatus = 0
-	DomainWarning           DomainStatus = 1
-	DomainFailure           DomainStatus = 2
-	DomainError             DomainStatus = 3
-	DomainNoSTARTTLSFailure DomainStatus = 4
-	DomainCouldNotConnect   DomainStatus = 5
+	DomainSuccess            DomainStatus = 0
+	DomainWarning            DomainStatus = 1
+	DomainFailure            DomainStatus = 2
+	DomainError              DomainStatus = 3
+	DomainNoSTARTTLSFailure  DomainStatus = 4
+	DomainCouldNotConnect    DomainStatus = 5
+	DomainBadHostnameFailure DomainStatus = 6
 )
 
+// This interface allows us to mock the implementation
+// of the actual checker.
+type domainCheckQuery interface {
+	// Performs an MX lookup to resolve the hostnames for this domain.
+	lookupHostnames() ([]string, error)
+	// Performs a subcheck on the given hostname.
+	checkHostname(string) HostnameResult
+	// Returns the FQDN that we're checking.
+	getDomain() string
+	// Returns the hostname patterns that we expect.
+	getExpectedHostnames() []string
+}
+
+// DomainQuery wraps the parameters we need to perform a domain check.
+type DomainQuery struct {
+	// Domain being checked.
+	Domain string
+	// Expected hostnames in MX records for Domain
+	ExpectedHostnames []string
+}
```

Comment:
I wonder if it might be preferable to get rid of the `domainCheckQuery` interface and just wrap the functions that require mocks in interfaces - something like this:

```
type DomainQuery struct {
	// Domain being checked.
	Domain string
	// Expected hostnames in MX records for Domain
	ExpectedHostnames []string
	hostnameLookup
	hostnameChecker
}

type hostnameLookup interface {
	lookupHostnames() []string
}

type hostnameChecker interface {
	checkHostname(string) HostnameResult
}
```

That would make the mocked functions explicit and we wouldn't need to add getters or wrappers for any additional functions we might want to call from the tests that don't request mocks. I didn't actually code it out, so I'm not sure it's actually better - was just playing a bit to see if the mocks could be leaner. Mostly influenced by the [code sample](https://github.com/sendgrid/filegetter/blob/master/getter/getter.go) that came along with this [blog post](https://sendgrid.com/blog/when-writing-unit-tests-dont-use-mocks/).

Reply:
I like this better-- thanks!! tried it out in most recent commit; lmk what you think!